libs/header/p2p/subscription.go

```diff
@@ -0,0 +1,54 @@
+package p2p
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+
+	"github.com/celestiaorg/celestia-node/libs/header"
+)
+
+// subscription handles retrieving Headers from the header pubsub topic.
+type subscription[H header.Header] struct {
+	topic        *pubsub.Topic
+	subscription *pubsub.Subscription
+}
+
+// newSubscription creates a new Header event subscription
+// on the given host.
+func newSubscription[H header.Header](topic *pubsub.Topic) (*subscription[H], error) {
+	sub, err := topic.Subscribe()
+	if err != nil {
+		return nil, err
+	}
+
+	return &subscription[H]{
+		topic:        topic,
+		subscription: sub,
+	}, nil
+}
+
+// NextHeader returns the next (latest) verified Header from the network.
+func (s *subscription[H]) NextHeader(ctx context.Context) (H, error) {
+	msg, err := s.subscription.Next(ctx)
+	if err != nil {
+		var zero H
+		return zero, err
+	}
+	log.Debugw("received message", "topic", msg.Message.GetTopic(), "sender", msg.ReceivedFrom)
+
+	header, ok := msg.ValidatorData.(H)
+	if !ok {
+		panic(fmt.Sprintf("invalid type received %s", reflect.TypeOf(msg.ValidatorData)))
```

Comment:
probably shouldn't be a panic? - wouldn't it be possible to send a header of a diff type if you're connected to the specific protocolID of the sub?

Reply:
This code is moved from header/p2p/subscription (again, Github has difficulties to show this properly).

I don't feel competent to answer your question.

CC: @Wondertan @vgonkivs

Reply:
It should be a panic. Otherwise, there is a critical bug in the code. We set H type to Validator Data explicitly in AddValidator